spring-rabbit/src/test/java/org/springframework/amqp/rabbit/listener/BrokerDeclaredQueueNameTests.java

```diff
@@ -0,0 +1,200 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.amqp.rabbit.listener;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.junit.jupiter.api.Test;
+
+import org.springframework.amqp.core.Message;
+import org.springframework.amqp.core.Queue;
+import org.springframework.amqp.rabbit.connection.CachingConnectionFactory;
+import org.springframework.amqp.rabbit.core.RabbitAdmin;
+import org.springframework.amqp.rabbit.core.RabbitTemplate;
+import org.springframework.amqp.rabbit.junit.RabbitAvailable;
+import org.springframework.amqp.rabbit.junit.RabbitAvailableCondition;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.test.annotation.DirtiesContext;
+import org.springframework.test.context.junit.jupiter.SpringJUnitConfig;
+
+/**
+ * @author Gary Russell
+ * @since 2.0.4.
+ *
+ */
+@SpringJUnitConfig
+@RabbitAvailable
+@DirtiesContext
+public class BrokerDeclaredQueueNameTests {
+
+	@Autowired
+	private CachingConnectionFactory cf;
+
+	@Autowired
+	private CountDownLatch latch1;
+
+	@Autowired
+	private CountDownLatch latch2;
+
+	@Autowired
+	private CountDownLatch latch3;
+
+	@Autowired
+	private CountDownLatch latch4;
+
+	@Autowired
+	private Queue queue1;
+
+	@Autowired
+	private Queue queue2;
+
+	@Autowired
+	private AtomicReference<Message> message;
+
+	@Autowired
+	private RabbitTemplate template;
+
+	@Autowired
+	private SimpleMessageListenerContainer smlc;
+
+	@Autowired
+	private DirectMessageListenerContainer dmlc;
+
+	@Test
+	public void testBrokerNamedQueueSMLC() throws Exception {
+		testBrokerNamedQueue(this.smlc, this.latch1, this.latch2, this.queue1);
+	}
+
+	@Test
+	public void testBrokerNamedQueueDMLC() throws Exception {
+		testBrokerNamedQueue(this.dmlc, this.latch3, this.latch4, this.queue2);
+	}
+
+	private void testBrokerNamedQueue(AbstractMessageListenerContainer container,
+			CountDownLatch latch1, CountDownLatch latch2, Queue queue) throws Exception {
+		container.start();
+		String firstDeclaredName = queue.getDeclaredName();
+		this.template.convertAndSend(firstDeclaredName, "foo");
+		assertThat(latch1.await(10, TimeUnit.SECONDS)).isTrue();
+		assertThat(this.message.get().getBody()).isEqualTo("foo".getBytes());
```

Comment:
You don't reset this `AtomicReference` in between tests: some stale data may remain.